### ISSUE TITLE
[getter-return] Extend ESLint's getter return to better fit TypeScript (second try)

### ISSUE THREAD

**Issue by vkrol** [typescript-eslint/typescript-eslint#3651]
Hi!

I apologize for the revival of the closed issue https://github.com/typescript-eslint/typescript-eslint/issues/2897, but I have an important observation in my opinion :)

> So using the getter-return lint rule is entirely superfluous and duplicates the TS errors.

This statement is not completely correct, because this rule also checks the getter when declaring a property with `Object.defineProperty`, while TS does not do this.

ESLint
```ts
Object.defineProperty({}, "foo", {
  get() {} // Expected to return a value in method 'get'. (getter-return)
});
```
https://eslint.org/demo#eyJ0ZXh0IjoiT2JqZWN0LmRlZmluZVByb3BlcnR5KHt9LCBcImZvb1wiLCB7XG4gIGdldCgpIHt9XG59KTsiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjEyLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7ImdldHRlci1yZXR1cm4iOjJ9LCJlbnYiOnt9fX0=

TypeScript
```ts
Object.defineProperty({}, "foo", {
  get() {} // No error
});
```
https://www.typescriptlang.org/play?#code/PIIwVgpgxgLgdAEwgMwJYDsIAUBOB7ABwhxgE8AKAbwF8AaAAgCJk89GHKAoe+gcwhjkAlPRr0A9OPoA5PPWL4cnakIDcQA

Therefore, disabling this rule for TS code can lead to the worsening of the static analysis quality.

I think a extension rule for `getter-return` should exist and should check for `return` in `Object.defineProperty` and do nothing with regular getters.

**Additional Info**

Nothing.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.28.0` |
| `@typescript-eslint/parser`        | `4.28.0` |
| `TypeScript`                       | `4.3.5` |
| `ESLint`                           | `7.31.0` |
| `node`                             | `16.5.0` |

**Comment by bradzacher** [comment]
I honestly don't think it's worth the effort for two reasons

First:
defining properties on things using `Object.defineProperty` is already SUPER rare of a usecase. The vast, vast majority of code will never even need to consider touching the API.
This is doubly true in TS code because it has zero type safety.

Second:
In order for you to specifically need the extension rule - you have to be using logic which requires types to verify it has complete branch coverage.

I just don't think it's worth the effort implementing this as a type-aware lint rule which attempts to solve for this case. It wouldn't be a small amount of work, and it would be complicated to build.

It would be much easier and likely even better if this one case was handled by TS.

**Comment by vkrol** [comment]
> I just don't think it's worth the effort implementing this as a type-aware lint rule which attempts to solve for this case. It wouldn't be a small amount of work, and it would be complicated to build. That's a lot of volunteer effort to solve for an 0.001% edge case.

I totally agree with you. I don't mind closing this issue with `wontfix` label. I just wanted to keep this observation for history. Perhaps it will be useful to someone in the future.

> But also - can't this just be solved by annotating the return type on the get function?

Yes, sure.

---

Thank you so much for such a detailed response. And thank you also for your invaluable work on this project.
